nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecordTest.java

```diff
@@ -0,0 +1,850 @@
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESSFUL_RECORDS, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_ERROR_RESPONSES, 0);
+    }
+
+    @Test
+    public void testInputRequired() {
+        runner.run();
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESS, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILURE, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_RETRY, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILED_RECORDS, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESSFUL_RECORDS, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_ERROR_RESPONSES, 0);
+    }
+
+    @Test
+    public void testErrorRelationship() throws Exception {
+        JsonRecordSetWriter writer = new JsonRecordSetWriter();
+        runner.addControllerService("writer", writer);
+        runner.setProperty(writer, SchemaAccessUtils.SCHEMA_ACCESS_STRATEGY, SchemaAccessUtils.SCHEMA_NAME_PROPERTY);
+        runner.setProperty(writer, SchemaAccessUtils.SCHEMA_REGISTRY, "registry");
+        runner.enableControllerService(writer);
+        runner.setProperty(PutElasticsearchRecord.RESULT_RECORD_WRITER, "writer");
+        runner.setProperty(PutElasticsearchRecord.NOT_FOUND_IS_SUCCESSFUL, "true");
+        clientService.setResponse(IndexOperationResponse.fromJsonResponse(Files.readString(Paths.get(TEST_COMMON_DIR, "sampleErrorResponse.json"))));
+        registry.addSchema(ERROR_TEST_SCHEMA, errorTestSchema);
+        runner.enqueue(VALUES, Collections.singletonMap("schema.name", ERROR_TEST_SCHEMA));
+        runner.setProperty(PutElasticsearchRecord.LOG_ERROR_RESPONSES, "true");
+        runner.assertValid();
+        runner.run();
+
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESS, 1);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILURE, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_RETRY, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILED_RECORDS, 1);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESSFUL_RECORDS, 1);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_ERROR_RESPONSES, 0);
+
+        runner.getFlowFilesForRelationship(PutElasticsearchRecord.REL_FAILED_RECORDS).get(0).assertAttributeEquals(PutElasticsearchRecord.ATTR_RECORD_COUNT, "1");
+        runner.getFlowFilesForRelationship(PutElasticsearchRecord.REL_SUCCESSFUL_RECORDS).get(0).assertAttributeEquals(PutElasticsearchRecord.ATTR_RECORD_COUNT, "4");
+
+        assertEquals(1,
+                runner.getProvenanceEvents().stream().filter(e -> ProvenanceEventType.SEND.equals(e.getEventType())
+                        && "1 Elasticsearch _bulk operation batch(es) [1 error(s), 4 success(es)]".equals(e.getDetails())).count());
+    }
+
+    @Test
+    public void testErrorsLoggedWithoutErrorRelationship() throws Exception {
```

Comment:
It would be better to rationalise these test cases and just have the differences present in individual test cases, e.g.

```java
private void testErrors(final int failedRecordCount, final int successfulRecordCount, final int errorResponseCount) throws Exception {
        final JsonRecordSetWriter writer = new JsonRecordSetWriter();
        runner.addControllerService("writer", writer);
        runner.setProperty(writer, SchemaAccessUtils.SCHEMA_ACCESS_STRATEGY, SchemaAccessUtils.SCHEMA_NAME_PROPERTY);
        runner.setProperty(writer, SchemaAccessUtils.SCHEMA_REGISTRY, "registry");
        runner.enableControllerService(writer);
        clientService.setResponse(IndexOperationResponse.fromJsonResponse(Files.readString(Paths.get(TEST_COMMON_DIR, "sampleErrorResponse.json"))));
        registry.addSchema(ERROR_TEST_SCHEMA, errorTestSchema);
        runner.setProperty(PutElasticsearchRecord.RESULT_RECORD_WRITER, "writer");

        runner.assertValid();

        runner.enqueue(VALUES, Collections.singletonMap("schema.name", ERROR_TEST_SCHEMA));
        runner.run();

        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESS, 1);
        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILURE, 0);
        runner.assertTransferCount(PutElasticsearchRecord.REL_RETRY, 0);
        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILED_RECORDS, failedRecordCount > 0 ? 1 : 0);
        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESSFUL_RECORDS, successfulRecordCount > 0 ? 1 : 0);
        runner.assertTransferCount(PutElasticsearchRecord.REL_ERROR_RESPONSES, errorResponseCount > 0 ? 1 : 0);

        if (failedRecordCount > 0) {
            runner.getFlowFilesForRelationship(PutElasticsearchRecord.REL_FAILED_RECORDS).get(0).assertAttributeEquals(PutElasticsearchRecord.ATTR_RECORD_COUNT, String.valueOf(failedRecordCount));
        }
        if (successfulRecordCount > 0) {
            runner.getFlowFilesForRelationship(PutElasticsearchRecord.REL_SUCCESSFUL_RECORDS).get(0).assertAttributeEquals(PutElasticsearchRecord.ATTR_RECORD_COUNT, String.valueOf(successfulRecordCount));
        }
        if (errorResponseCount > 0) {
            final String errorResponses = runner.getFlowFilesForRelationship(PutElasticsearchJson.REL_ERROR_RESPONSES).get(0).getContent();
            assertTrue(errorResponses.contains("not_found"));
            assertTrue(errorResponses.contains("For input string: 20abc"));
        }

        assertEquals(1,
                runner.getProvenanceEvents().stream().filter(e -> ProvenanceEventType.SEND.equals(e.getEventType())
                        && "1 Elasticsearch _bulk operation batch(es) [%d error(s), %d success(es)]".formatted(failedRecordCount, successfulRecordCount).equals(e.getDetails())).count());
    }

    @Test
    public void testErrorRelationship() throws Exception {
        runner.setProperty(PutElasticsearchRecord.NOT_FOUND_IS_SUCCESSFUL, "true");
        runner.setProperty(PutElasticsearchRecord.LOG_ERROR_RESPONSES, "true");
        testErrors(1, 4, 0);
    }

    @Test
    public void testErrorsLoggedWithoutErrorRelationship() throws Exception {
        runner.setProperty(PutElasticsearchRecord.NOT_FOUND_IS_SUCCESSFUL, "false");
        testErrors(2, 3, 0);
    }

    @Test
    public void testDocumentNotFoundErrorHandling() throws Exception {
        runner.setProperty(PutElasticsearchRecord.NOT_FOUND_IS_SUCCESSFUL, "false");
        runner.setProperty(PutElasticsearchRecord.OUTPUT_ERROR_RESPONSES, "true");
        testErrors(2, 3, 1);
    }
```

This will reduce duplicated code across a few of the test cases

Reply:
I do not see the following block currently in `testDocumentNotFoundErrorHandling`

 ```
final JsonRecordSetWriter writer = new JsonRecordSetWriter();
        runner.addControllerService("writer", writer);
        runner.setProperty(writer, SchemaAccessUtils.SCHEMA_ACCESS_STRATEGY, SchemaAccessUtils.SCHEMA_NAME_PROPERTY);
        runner.setProperty(writer, SchemaAccessUtils.SCHEMA_REGISTRY, "registry");
        runner.enableControllerService(writer);
        clientService.setResponse(IndexOperationResponse.fromJsonResponse(Files.readString(Paths.get(TEST_COMMON_DIR, "sampleErrorResponse.json"))));
        registry.addSchema(ERROR_TEST_SCHEMA, errorTestSchema);
        runner.setProperty(PutElasticsearchRecord.RESULT_RECORD_WRITER, "writer");
```

can it still be used for it?

Reply:
NM I see that it does not make a difference